File: src/types/provider.ts
export type ApiType =
  | "openai-completions"
  | "openai-responses"
  | "openai-codex-responses"
  | "azure-openai-responses"
  | "anthropic-messages"
  | "google-generative-ai"
  | "google-vertex";

export interface ProviderConfig {
  id: string;
  name: string;
  enabled: boolean;
  isBuiltIn: boolean;
  baseUrl?: string;
  apiKey?: string;
  api?: ApiType;
  headers?: Record<string, string>;
  authHeader?: boolean;
  auth?: "apiKey" | "none";
  discovery?: { type: "ollama" | "lm-studio" };
  modelOverrides?: Record<string, ModelOverride>;
  models?: ModelDefinition[];
  createdAt?: string;
  updatedAt?: string;
}

export interface ModelDefinition {
  id: string;
  name: string;
  api?: ApiType;
  reasoning: boolean;
  input: ("text" | "image")[];
  cost: {
    input: number;
    output: number;
    cacheRead: number;
    cacheWrite: number;
    contextOver200k?: {
      input: number;
      output: number;
      cacheRead?: number;
      cacheWrite?: number;
    };
  };
  contextWindow: number;
  maxTokens: number;
  headers?: Record<string, string>;
  compat?: ModelCompat;
  // OpenAI protocol parameter defaults
  defaultTemperature?: number;
  defaultTopP?: number;
  defaultPresencePenalty?: number;
  defaultFrequencyPenalty?: number;
  defaultSeed?: number;
  // New fields from PRD
  family?: string;
  releaseDate?: string;
  attachment?: boolean;
  temperature?: boolean;
  toolCall?: boolean;
  interleaved?: true | { field: "reasoning_content" | "reasoning_details" };
  status?: "alpha" | "beta" | "deprecated";
  provider?: { npm?: string; api?: string };
  modalities?: {
    input?: string[];
    output?: string[];
  };
  limit?: {
    context?: number;
    input?: number;
    output?: number;
  };
  thinkingLevelMap?: Record<string, string | null>;
  options?: Record<string, unknown>;
  variants?: Record<string, { disabled?: boolean } & Record<string, unknown>>;
  experimental?: boolean | {
    modes?: Record<string, {
      cost?: { input: number; output: number; cacheRead?: number; cacheWrite?: number };
      provider?: { body?: Record<string, unknown>; headers?: Record<string, string> };
    }>;
  };
}

export interface ModelOverride {
  name?: string;
  reasoning?: boolean;
  input?: ("text" | "image")[];
  cost?: {
    input: number;
    output: number;
    cacheRead: number;
    cacheWrite: number;
    contextOver200k?: {
      input: number;
      output: number;
      cacheRead?: number;
      cacheWrite?: number;
    };
  };
  contextWindow?: number;
  maxTokens?: number;
  headers?: Record<string, string>;
  compat?: ModelCompat;
  // OpenAI protocol parameter defaults
  defaultTemperature?: number;
  defaultTopP?: number;
  defaultPresencePenalty?: number;
  defaultFrequencyPenalty?: number;
  defaultSeed?: number;
  contextPromotionTarget?: string;
  family?: string;
  releaseDate?: string;
  attachment?: boolean;
  temperature?: boolean;
  toolCall?: boolean;
  interleaved?: true | { field: "reasoning_content" | "reasoning_details" };
  status?: "alpha" | "beta" | "deprecated";
  modalities?: {
    input?: string[];
    output?: string[];
  };
  limit?: {
    context?: number;
    input?: number;
    output?: number;
  };
  thinkingLevelMap?: Record<string, string | null>;
  options?: Record<string, unknown>;
  variants?: Record<string, { disabled?: boolean } & Record<string, unknown>>;
  experimental?: boolean | {
    modes?: Record<string, {
      cost?: { input: number; output: number; cacheRead?: number; cacheWrite?: number };
      provider?: { body?: Record<string, unknown>; headers?: Record<string, string> };
    }>;
  };
}

export interface ModelCompat {
  supportsStore?: boolean;
  supportsDeveloperRole?: boolean;
  supportsReasoningEffort?: boolean;
  maxTokensField?: string;
  openRouterRouting?: Record<string, unknown>;
  vercelGatewayRouting?: Record<string, unknown>;
  extraBody?: Record<string, unknown>;
  // New compat fields from PRD
  supportsUsageInStreaming?: boolean;
  requiresToolResultName?: boolean;
  requiresAssistantAfterToolResult?: boolean;
  requiresThinkingAsText?: boolean;
  requiresReasoningContentOnAssistantMessages?: boolean;
  thinkingFormat?: string;
  cacheControlFormat?: string;
  supportsStrictMode?: boolean;
  supportsLongCacheRetention?: boolean;
  supportsEagerToolInputStreaming?: boolean;
  forceAdaptiveThinking?: boolean;
  allowEmptySignature?: boolean;
  supportsCacheControlOnTools?: boolean;
  sendSessionAffinityHeaders?: boolean;
}

File: src-tauri/src/models/provider.rs
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ProviderConfig {
    pub id: String,
    pub name: String,
    pub enabled: bool,
    pub is_built_in: bool,
    pub base_url: Option<String>,
    pub api_key: Option<String>,
    pub api_type: Option<String>,
    pub headers: Option<HashMap<String, String>>,
    pub auth_header: Option<bool>,
    pub auth: Option<String>,
    pub discovery: Option<DiscoveryConfig>,
    pub model_overrides: Option<HashMap<String, ModelOverride>>,
    pub models: Option<Vec<ModelDefinition>>,
    pub created_at: Option<String>,
    pub updated_at: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DiscoveryConfig {
    #[serde(rename = "type")]
    pub discovery_type: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ModelDefinition {
    pub id: String,
    pub name: String,
    pub api_type: Option<String>,
    pub reasoning: bool,
    pub input_types: Vec<String>,
    pub cost: ModelCost,
    pub context_window: i64,
    pub max_tokens: i64,
    pub headers: Option<HashMap<String, String>>,
    pub compat: Option<ModelCompat>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ModelCost {
    pub input: f64,
    pub output: f64,
    #[serde(rename = "cacheRead")]
    pub cache_read: f64,
    #[serde(rename = "cacheWrite")]
    pub cache_write: f64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ModelCompat {
    #[serde(rename = "supportsStore")]
    pub supports_store: Option<bool>,
    #[serde(rename = "supportsDeveloperRole")]
    pub supports_developer_role: Option<bool>,
    #[serde(rename = "supportsReasoningEffort")]
    pub supports_reasoning_effort: Option<bool>,
    #[serde(rename = "maxTokensField")]
    pub max_tokens_field: Option<String>,
    #[serde(rename = "openRouterRouting")]
    pub open_router_routing: Option<serde_json::Value>,
    #[serde(rename = "vercelGatewayRouting")]
    pub vercel_gateway_routing: Option<serde_json::Value>,
    #[serde(rename = "extraBody")]
    pub extra_body: Option<serde_json::Value>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ModelOverride {
    pub name: Option<String>,
    pub reasoning: Option<bool>,
    pub input: Option<Vec<String>>,
    pub cost: Option<ModelCost>,
    #[serde(rename = "contextWindow")]
    pub context_window: Option<i64>,
    #[serde(rename = "maxTokens")]
    pub max_tokens: Option<i64>,
    pub headers: Option<HashMap<String, String>>,
    pub compat: Option<ModelCompat>,
    #[serde(rename = "contextPromotionTarget")]
    pub context_promotion_target: Option<String>,
}
